ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,334 @@
+              return new ThreadSaturation(threadName, startTime, window, sampleMargin);
+            }
+        );
+        LOGGER.debug("Record thread {} sample {} {}", threadName, totalBlocked, startTime);
+        final BlockedTimeSample blockedTimeSample = new BlockedTimeSample(now, totalBlocked);
+        final Optional<Double> measured = threadSaturation.measure(now, blockedTimeSample);
+        LOGGER.debug(
+            "Measured value for thread {}: {}",
+            threadName,
+            (measured.map(Object::toString).orElse(""))
+        );
+        measured.ifPresent(m -> reportThreadSaturation(now, m, threadName, reporter));
+        saturation = compareSaturation(saturation, measured) > 0 ? saturation : measured;
+      }
+      saturation.ifPresent(s -> reportQuerySaturation(now, s, queryId, reporter));
+      for (final String threadName
+          : Sets.difference(new HashSet<>(perThreadSaturation.keySet()), byThread.keySet())) {
+        perThreadSaturation.remove(threadName);
+        reporter.cleanup(QUERY_THREAD_SATURATION, ImmutableMap.of(THREAD_ID, threadName));
+      }
+      return saturation;
+    }
+
+    private void cleanup(final MetricsReporter reporter) {
+      for (final String threadName : perThreadSaturation.keySet()) {
+        reporter.cleanup(QUERY_THREAD_SATURATION, ImmutableMap.of(THREAD_ID, threadName));
+      }
+      reporter.cleanup(QUERY_SATURATION, ImmutableMap.of(QUERY_ID, queryId.toString()));
+    }
+  }
+
+  private static final class ThreadSaturation {
+    private final String threadName;
+    private final List<BlockedTimeSample> samples = new LinkedList<>();
+    private final Instant startTime;
+    private final Duration window;
+    private final Duration sampleMargin;
+
+    private ThreadSaturation(
+        final String threadName,
+        final long startTime,
+        final Duration window,
+        final Duration sampleMargin
+    ) {
+      this.threadName = Objects.requireNonNull(threadName, "threadName");
+      this.startTime = Instant.ofEpochMilli(startTime);
+      this.window = Objects.requireNonNull(window, "window");
+      this.sampleMargin = Objects.requireNonNull(sampleMargin, "sampleMargin");
+    }
+
+    private boolean inRange(final Instant time, final Instant start, final Instant end) {
+      return time.isAfter(start) && time.isBefore(end);
+    }
+
+    private Optional<Double> measure(final Instant now, final BlockedTimeSample current) {
+      final Instant windowStart = now.minus(window);
+      final Instant earliest = now.minus(window.plus(sampleMargin));
+      final Instant latest = now.minus(window.minus(sampleMargin));
+      LOGGER.debug("{}: record and measure with now {},  window {} ({} : {})", threadName, now, windowStart, earliest, latest);
+      samples.add(current);
+      samples.removeIf(s -> s.timestamp.isBefore(earliest));
+      if (!inRange(samples.get(0).timestamp, earliest, latest) && !startTime.isAfter(windowStart)) {
```

Comment:
> Why is it okay for the metric to be out of range if the thread was started after windowStart?
If the thread was started after `windowStart` we want to consider the time from `windowStart` to the thread start time as blocked.

>  I guess since we're no longer able to guarantee that the sample time happens every 5 minutes this helps us get a good set of metrics for our window?

Yep.

>  If so, why is sampleMargin a variable instead of a fixed time we decide on?

I wanted to leave the option open to have it be configurable. But it shouldn't ever change once the server is started.